packages/jni-swig-stub/src/main/jni/realm_api_helpers.cpp

```diff
@@ -664,19 +664,25 @@ before_client_reset(void* userdata, realm_t* before_realm) {
     auto before_pointer = wrap_pointer(env, reinterpret_cast<jlong>(before_realm), false);
     env->CallVoidMethod(static_cast<jobject>(userdata), java_before_callback_function, before_pointer);
     jni_check_exception(env);
+
+    return true;
```

Comment:
I couldn't any changes to code on the JVM side, which means that if the callback throws on the JVM, then `jni_check_exception(env);` will throw an exception instead of returning false.

Do we have any unit tests of us manually throwing an exception in `BeforeClientReset`? I

Reply:
Yes, `onBefore` here https://github.com/realm/realm-kotlin/pull/967/files#diff-4496cd276f4c265c4a34d2d6835c8a1e290376d9d2d30380bb4dc60ac49a94ebR547
and `onAfter` here https://github.com/realm/realm-kotlin/pull/967/files#diff-4496cd276f4c265c4a34d2d6835c8a1e290376d9d2d30380bb4dc60ac49a94ebR599

Reply:
I debugged it when I wrote this for automatic client reset in a separate branch and looks like `jni_check_exception` does the catching and forwards the error appropriately.

Reply:
After carefully inspecting the execution it seems it does work because the error is intercepted by the error handler instead, then we detect the error itself has `isClientResetRequested` so it's routed through the client reset handler. I will change it so that it doesn't take that path and stays within the client reset flow.

Reply:
Funnily enough the before/after callbacks in the C-API throw an exception too:
```
        if (!callback(userdata.get(), &r1)) {
            throw CallbackFailed();
        }
```
so the execution stays the same regardless. I will `return false` for consistency with native though.

Reply:
The exception catch could be done in the cinterop with a lambda.